[Java] SetFromMapSerializer didn't handle nested reference right

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/fury/issues) and found no similar issues.


### Version

main

### Component(s)

Java

### Minimal reproduce step
```java
  @Data
  @AllArgsConstructor
  public static class CollectionViewTestStruct {
    Collection<String> collection;
    Set<String> set;
  }

  @Test(dataProvider = "javaFury")
  public void testSetFromMap(Fury fury) {
    Set<String> set = Collections.newSetFromMap(Maps.newConcurrentMap());
    set.add("a");
    set.add("b");
    set.add("c");
    serDeCheck(fury, set);
    Assert.assertEquals(
      fury.getClassResolver().getSerializerClass(set.getClass()),
        CollectionSerializers.SetFromMapSerializer.class);
    serDeCheck(fury, new CollectionViewTestStruct(set, set));
    set = Collections.newSetFromMap(new HashMap<String, Boolean>() {});
    set.add("a");
    set.add("b");
    serDeCheck(fury, set);
    serDeCheck(fury, new CollectionViewTestStruct(set, set));
  }
```

### What did you expect to see?

//

### What did you see instead?

```java
org.apache.fury.exception.DeserializationException: Deserialize failed, read objects are: [{}]

	at org.apache.fury.util.ExceptionUtils.handleReadFailed(ExceptionUtils.java:63)
	at org.apache.fury.Fury.deserialize(Fury.java:796)
	at org.apache.fury.Fury.deserialize(Fury.java:714)
	at org.apache.fury.FuryTestBase.serDe(FuryTestBase.java:270)
	at org.apache.fury.FuryTestBase.serDeCheck(FuryTestBase.java:262)
	at org.apache.fury.serializer.collection.CollectionSerializersTest.testSetFromMap(CollectionSerializersTest.java:409)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:658)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:219)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:923)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:192)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at org.testng.TestRunner.privateRun(TestRunner.java:808)
	at org.testng.TestRunner.run(TestRunner.java:603)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:429)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:423)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:383)
	at org.testng.SuiteRunner.run(SuiteRunner.java:326)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1249)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
	at org.testng.TestNG.runSuites(TestNG.java:1092)
	at org.testng.TestNG.run(TestNG.java:1060)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:105)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 4
	at org.apache.fury.collection.IntArray.pop(IntArray.java:54)
	at org.apache.fury.resolver.MapRefResolver.reference(MapRefResolver.java:192)
	at org.apache.fury.serializer.collection.CollectionSerializers$SetFromMapSerializer.newCollection(CollectionSerializers.java:482)
	at org.apache.fury.serializer.collection.CollectionSerializer.read(CollectionSerializer.java:68)
	at org.apache.fury.serializer.collection.CollectionSerializer.read(CollectionSerializer.java:27)
	at org.apache.fury.Fury.readDataInternal(Fury.java:958)
	at org.apache.fury.Fury.readRef(Fury.java:860)
	at org.apache.fury.Fury.deserialize(Fury.java:792)
	... 29 more
```

### Anything Else?

https://github.com/apache/fury/pull/1616

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractCollectionSerializer.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractCollectionSerializer.java
@@ -514,9 +514,7 @@ public abstract class AbstractCollectionSerializer<T> extends Serializer<T> {
     }
   }
 
-  /**
-   * Create a new empty collection for copy.
-   */
+  /** Create a new empty collection for copy. */
   public Collection newCollection(Collection collection) {
     numElements = collection.size();
     if (constructor == null) {

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractCollectionSerializer.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractCollectionSerializer.java
@@ -41,7 +41,7 @@ import org.apache.fury.util.Preconditions;
 public abstract class AbstractCollectionSerializer<T> extends Serializer<T> {
   private MethodHandle constructor;
   private int numElements;
-  private final boolean supportCodegenHook;
+  protected final boolean supportCodegenHook;
   // TODO remove elemSerializer, support generics in CompatibleSerializer.
   private Serializer<?> elemSerializer;
   protected final ClassInfoHolder elementClassInfoHolder;
@@ -514,23 +514,11 @@ public abstract class AbstractCollectionSerializer<T> extends Serializer<T> {
     }
   }
 
+  /**
+   * Create a new empty collection for copy.
+   */
   public Collection newCollection(Collection collection) {
     numElements = collection.size();
-    return newCollection();
-  }
-
-  /**
-   * Collection must have default constructor to be invoked by fury, otherwise created object can't
-   * be used to adding elements. For example:
-   *
-   * <pre>{@code new ArrayList<Integer> {add(1);}}</pre>
-   *
-   * <p>without default constructor, created list will have elementData as null, adding elements
-   * will raise NPE.
-   *
-   * @return empty collection instance
-   */
-  public Collection newCollection() {
     if (constructor == null) {
       constructor = ReflectionUtils.getCtrHandle(type, true);
     }

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractMapSerializer.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractMapSerializer.java
@@ -800,9 +800,7 @@ public abstract class AbstractMapSerializer<T> extends Serializer<T> {
     }
   }
 
-  /**
-   * Create a new empty map for copy.
-   */
+  /** Create a new empty map for copy. */
   public Map newMap(Map map) {
     numElements = map.size();
     if (constructor == null) {
@@ -812,7 +810,7 @@ public abstract class AbstractMapSerializer<T> extends Serializer<T> {
       return (Map) constructor.invoke();
     } catch (Throwable e) {
       throw new IllegalArgumentException(
-        "Please provide public no arguments constructor for class " + type, e);
+          "Please provide public no arguments constructor for class " + type, e);
     }
   }
 

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractMapSerializer.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/AbstractMapSerializer.java
@@ -800,23 +800,11 @@ public abstract class AbstractMapSerializer<T> extends Serializer<T> {
     }
   }
 
+  /**
+   * Create a new empty map for copy.
+   */
   public Map newMap(Map map) {
     numElements = map.size();
-    return newMap();
-  }
-
-  /**
-   * Map must have default constructor to be invoked by fury, otherwise created object can't be used
-   * to adding elements. For example:
-   *
-   * <pre>{@code new ArrayList<Integer> {add(1);}}</pre>
-   *
-   * <p>without default constructor, created list will have elementData as null, adding elements
-   * will raise NPE.
-   *
-   * @return empty map instance
-   */
-  public Map newMap() {
     if (constructor == null) {
       constructor = ReflectionUtils.getCtrHandle(type, true);
     }
@@ -824,7 +812,7 @@ public abstract class AbstractMapSerializer<T> extends Serializer<T> {
       return (Map) constructor.invoke();
     } catch (Throwable e) {
       throw new IllegalArgumentException(
-          "Please provide public no arguments constructor for class " + type, e);
+        "Please provide public no arguments constructor for class " + type, e);
     }
   }
 

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializer.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializer.java
@@ -22,6 +22,7 @@ package org.apache.fury.serializer.collection;
 import java.util.Collection;
 import org.apache.fury.Fury;
 import org.apache.fury.memory.MemoryBuffer;
+import org.apache.fury.util.Preconditions;
 
 /** Base serializer for all java collections. */
 @SuppressWarnings({"unchecked", "rawtypes"})
@@ -55,6 +56,7 @@ public class CollectionSerializer<T extends Collection> extends AbstractCollecti
     if (isImmutable()) {
       return originCollection;
     }
+    Preconditions.checkArgument(supportCodegenHook);
     Collection newCollection = newCollection(originCollection);
     if (needToCopyRef) {
       fury.reference(originCollection, newCollection);

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
@@ -461,11 +461,15 @@ public class CollectionSerializers {
   public static final class SetFromMapSerializer extends CollectionSerializer<Set<?>> {
 
     private static final long MAP_FIELD_OFFSET;
+    private static final MethodHandle ctr;
+    private static final List EMPTY_COLLECTION_STUB = new ArrayList<>();
 
     static {
       try {
-        Field mapField = Class.forName("java.util.Collections$SetFromMap").getDeclaredField("m");
+        Class<?> type = Class.forName("java.util.Collections$SetFromMap");
+        Field mapField = type.getDeclaredField("m");
         MAP_FIELD_OFFSET = Platform.objectFieldOffset(mapField);
+        ctr = ReflectionUtils.getCtrHandle(type, Map.class);
       } catch (final Exception e) {
         throw new RuntimeException(e);
       }
@@ -483,16 +487,20 @@ public class CollectionSerializers {
       RefResolver refResolver = fury.getRefResolver();
       // It's possible that elements or nested fields has circular ref to set.
       int refId = refResolver.lastPreservedRefId();
-      Map map;
+      Set set;
       if (buffer.readBoolean()) {
         refResolver.preserveRefId();
-        map = mapSerializer.newMap(buffer);
+        set = Collections.newSetFromMap(mapSerializer.newMap(buffer));
         setNumElements(mapSerializer.getAndClearNumElements());
       } else {
-        map = (Map) mapSerializer.read(buffer);
+        Map map = (Map) mapSerializer.read(buffer);
+        try {
+          set = (Set) ctr.invokeWithArguments(map);
+        } catch (Throwable e) {
+          throw new RuntimeException(e);
+        }
         setNumElements(0);
       }
-      Set set = Collections.newSetFromMap(map);
       refResolver.setReadObject(refId, set);
       return set;
     }
@@ -516,11 +524,12 @@ public class CollectionSerializers {
       if (mapSerializer.supportCodegenHook) {
         buffer.writeBoolean(true);
         mapSerializer.onMapWrite(buffer, map);
+        return value;
       } else {
         buffer.writeBoolean(false);
         mapSerializer.write(buffer, map);
+        return EMPTY_COLLECTION_STUB;
       }
-      return value;
     }
   }
 
@@ -556,7 +565,8 @@ public class CollectionSerializers {
 
     @Override
     public Collection newCollection(MemoryBuffer buffer) {
-      throw new UnsupportedOperationException();
+      throw new IllegalStateException(
+          "Should not be invoked since we set supportCodegenHook to false");
     }
   }
 

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
@@ -533,12 +533,13 @@ public class CollectionSerializers {
     }
   }
 
-  public static final class ConcurrentHashMapKeySetView
+  public static final class ConcurrentHashMapKeySetViewSerializer
       extends CollectionSerializer<ConcurrentHashMap.KeySetView> {
     private final ClassInfoHolder mapClassInfoHolder;
     private final ClassInfoHolder valueClassInfoHolder;
 
-    public ConcurrentHashMapKeySetView(Fury fury, Class<ConcurrentHashMap.KeySetView> type) {
+    public ConcurrentHashMapKeySetViewSerializer(
+        Fury fury, Class<ConcurrentHashMap.KeySetView> type) {
       super(fury, type, false);
       mapClassInfoHolder = fury.getClassResolver().nilClassInfoHolder();
       valueClassInfoHolder = fury.getClassResolver().nilClassInfoHolder();
@@ -838,6 +839,6 @@ public class CollectionSerializers {
     fury.registerSerializer(setFromMapClass, new SetFromMapSerializer(fury, setFromMapClass));
     fury.registerSerializer(
         ConcurrentHashMap.KeySetView.class,
-        new ConcurrentHashMapKeySetView(fury, ConcurrentHashMap.KeySetView.class));
+        new ConcurrentHashMapKeySetViewSerializer(fury, ConcurrentHashMap.KeySetView.class));
   }
 }

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
@@ -20,6 +20,7 @@
 package org.apache.fury.serializer.collection;
 
 import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
 import java.lang.reflect.Field;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
@@ -59,6 +60,7 @@ import org.apache.fury.serializer.Serializer;
 import org.apache.fury.serializer.Serializers;
 import org.apache.fury.type.Type;
 import org.apache.fury.util.Preconditions;
+import org.apache.fury.util.unsafe._JDKAccess;
 
 /**
  * Serializers for classes implements {@link Collection}. All collection serializers should extend
@@ -461,15 +463,20 @@ public class CollectionSerializers {
   public static final class SetFromMapSerializer extends CollectionSerializer<Set<?>> {
 
     private static final long MAP_FIELD_OFFSET;
-    private static final MethodHandle ctr;
     private static final List EMPTY_COLLECTION_STUB = new ArrayList<>();
+
+    private static final MethodHandle m;
+
+    private static final MethodHandle s;
 
     static {
       try {
         Class<?> type = Class.forName("java.util.Collections$SetFromMap");
         Field mapField = type.getDeclaredField("m");
         MAP_FIELD_OFFSET = Platform.objectFieldOffset(mapField);
-        ctr = ReflectionUtils.getCtrHandle(type, Map.class);
+        MethodHandles.Lookup lookup = _JDKAccess._trustedLookup(type);
+        m = lookup.findSetter(type, "m", Map.class);
+        s = lookup.findSetter(type, "s", Set.class);
       } catch (final Exception e) {
         throw new RuntimeException(e);
       }
@@ -495,7 +502,9 @@ public class CollectionSerializers {
       } else {
         Map map = (Map) mapSerializer.read(buffer);
         try {
-          set = (Set) ctr.invokeWithArguments(map);
+          set = Platform.newInstance(type);
+          m.invoke(set, map);
+          s.invoke(set, map.keySet());
         } catch (Throwable e) {
           throw new RuntimeException(e);
         }

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
@@ -44,7 +44,6 @@ import java.util.Vector;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentSkipListSet;
 import java.util.concurrent.CopyOnWriteArrayList;
-
 import org.apache.fury.Fury;
 import org.apache.fury.config.Language;
 import org.apache.fury.exception.FuryException;
@@ -266,8 +265,7 @@ public class CollectionSerializers {
     }
 
     @Override
-    public void write(MemoryBuffer buffer, List<?> value) {
-    }
+    public void write(MemoryBuffer buffer, List<?> value) {}
 
     @Override
     public short getXtypeId() {
@@ -293,7 +291,7 @@ public class CollectionSerializers {
   }
 
   public static class CopyOnWriteArrayListSerializer
-    extends CollectionSerializer<CopyOnWriteArrayList> {
+      extends CollectionSerializer<CopyOnWriteArrayList> {
 
     public CopyOnWriteArrayListSerializer(Fury fury, Class<CopyOnWriteArrayList> type) {
       super(fury, type);
@@ -320,8 +318,7 @@ public class CollectionSerializers {
     }
 
     @Override
-    public void write(MemoryBuffer buffer, Set<?> value) {
-    }
+    public void write(MemoryBuffer buffer, Set<?> value) {}
 
     @Override
     public short getXtypeId() {
@@ -353,8 +350,7 @@ public class CollectionSerializers {
     }
 
     @Override
-    public void write(MemoryBuffer buffer, SortedSet<?> value) {
-    }
+    public void write(MemoryBuffer buffer, SortedSet<?> value) {}
 
     @Override
     public SortedSet<?> read(MemoryBuffer buffer) {
@@ -363,7 +359,7 @@ public class CollectionSerializers {
   }
 
   public static final class CollectionsSingletonListSerializer
-    extends CollectionSerializer<List<?>> {
+      extends CollectionSerializer<List<?>> {
 
     public CollectionsSingletonListSerializer(Fury fury, Class<List<?>> cls) {
       super(fury, cls, false);
@@ -442,7 +438,7 @@ public class CollectionSerializers {
   }
 
   public static final class ConcurrentSkipListSetSerializer
-    extends SortedSetSerializer<ConcurrentSkipListSet> {
+      extends SortedSetSerializer<ConcurrentSkipListSet> {
 
     public ConcurrentSkipListSetSerializer(Fury fury, Class<ConcurrentSkipListSet> cls) {
       super(fury, cls);
@@ -482,7 +478,8 @@ public class CollectionSerializers {
     @Override
     public Collection newCollection(MemoryBuffer buffer) {
       final ClassInfo mapClassInfo = fury.getClassResolver().readClassInfo(buffer);
-      final AbstractMapSerializer mapSerializer = (AbstractMapSerializer) mapClassInfo.getSerializer();
+      final AbstractMapSerializer mapSerializer =
+          (AbstractMapSerializer) mapClassInfo.getSerializer();
       RefResolver refResolver = fury.getRefResolver();
       // It's possible that elements or nested fields has circular ref to set.
       int refId = refResolver.lastPreservedRefId();
@@ -503,9 +500,9 @@ public class CollectionSerializers {
     @Override
     public Collection newCollection(Collection originCollection) {
       Map<?, Boolean> map =
-        (Map<?, Boolean>) Platform.getObject(originCollection, MAP_FIELD_OFFSET);
+          (Map<?, Boolean>) Platform.getObject(originCollection, MAP_FIELD_OFFSET);
       AbstractMapSerializer mapSerializer =
-        (AbstractMapSerializer) fury.getClassResolver().getSerializer(map.getClass());
+          (AbstractMapSerializer) fury.getClassResolver().getSerializer(map.getClass());
       Map newMap = mapSerializer.newMap(map);
       return Collections.newSetFromMap(newMap);
     }
@@ -528,7 +525,7 @@ public class CollectionSerializers {
   }
 
   public static final class ConcurrentHashMapKeySetView
-    extends CollectionSerializer<ConcurrentHashMap.KeySetView> {
+      extends CollectionSerializer<ConcurrentHashMap.KeySetView> {
     private final ClassInfoHolder mapClassInfoHolder;
     private final ClassInfoHolder valueClassInfoHolder;
 
@@ -649,7 +646,7 @@ public class CollectionSerializers {
     public void write(MemoryBuffer buffer, BitSet set) {
       long[] values = set.toLongArray();
       buffer.writePrimitiveArrayWithSize(
-        values, Platform.LONG_ARRAY_OFFSET, Math.multiplyExact(values.length, 8));
+          values, Platform.LONG_ARRAY_OFFSET, Math.multiplyExact(values.length, 8));
     }
 
     @Override
@@ -678,7 +675,7 @@ public class CollectionSerializers {
     @Override
     public Collection newCollection(Collection collection) {
       return new PriorityQueue(
-        collection.size(), fury.copyObject(((PriorityQueue) collection).comparator()));
+          collection.size(), fury.copyObject(((PriorityQueue) collection).comparator()));
     }
 
     @Override
@@ -698,19 +695,19 @@ public class CollectionSerializers {
    * but the correctness can be ensured.
    */
   public static final class DefaultJavaCollectionSerializer<T>
-    extends AbstractCollectionSerializer<T> {
+      extends AbstractCollectionSerializer<T> {
     private Serializer<T> dataSerializer;
 
     public DefaultJavaCollectionSerializer(Fury fury, Class<T> cls) {
       super(fury, cls, false);
       Preconditions.checkArgument(
-        fury.getLanguage() == Language.JAVA,
-        "Python default collection serializer should use " + CollectionSerializer.class);
+          fury.getLanguage() == Language.JAVA,
+          "Python default collection serializer should use " + CollectionSerializer.class);
       fury.getClassResolver().setSerializer(cls, this);
       Class<? extends Serializer> serializerClass =
-        fury.getClassResolver()
-          .getObjectSerializerClass(
-            cls, sc -> dataSerializer = Serializers.newSerializer(fury, cls, sc));
+          fury.getClassResolver()
+              .getObjectSerializerClass(
+                  cls, sc -> dataSerializer = Serializers.newSerializer(fury, cls, sc));
       dataSerializer = Serializers.newSerializer(fury, cls, serializerClass);
       // No need to set object serializer to this, it will be set in class resolver later.
       // fury.getClassResolver().setSerializer(cls, this);
@@ -742,11 +739,9 @@ public class CollectionSerializers {
     }
   }
 
-  /**
-   * Collection serializer for class with JDK custom serialization methods defined.
-   */
+  /** Collection serializer for class with JDK custom serialization methods defined. */
   public static final class JDKCompatibleCollectionSerializer<T>
-    extends AbstractCollectionSerializer<T> {
+      extends AbstractCollectionSerializer<T> {
     private final Serializer serializer;
 
     public JDKCompatibleCollectionSerializer(Fury fury, Class<T> cls) {
@@ -754,9 +749,9 @@ public class CollectionSerializers {
       // Collection which defined `writeReplace` may use this serializer, so check replace/resolve
       // is necessary.
       Class<? extends Serializer> serializerType =
-        ClassResolver.useReplaceResolveSerializer(cls)
-          ? ReplaceResolveSerializer.class
-          : fury.getDefaultJDKStreamSerializerType();
+          ClassResolver.useReplaceResolveSerializer(cls)
+              ? ReplaceResolveSerializer.class
+              : fury.getDefaultJDKStreamSerializerType();
       serializer = Serializers.newSerializer(fury, cls, serializerType);
     }
 
@@ -796,43 +791,43 @@ public class CollectionSerializers {
     Class arrayAsListClass = Arrays.asList(1, 2).getClass();
     fury.registerSerializer(arrayAsListClass, new ArraysAsListSerializer(fury, arrayAsListClass));
     fury.registerSerializer(
-      LinkedList.class, new CollectionSerializer(fury, LinkedList.class, true));
+        LinkedList.class, new CollectionSerializer(fury, LinkedList.class, true));
     fury.registerSerializer(HashSet.class, new HashSetSerializer(fury));
     fury.registerSerializer(LinkedHashSet.class, new LinkedHashSetSerializer(fury));
     fury.registerSerializer(TreeSet.class, new SortedSetSerializer<>(fury, TreeSet.class));
     fury.registerSerializer(
-      Collections.EMPTY_LIST.getClass(),
-      new EmptyListSerializer(fury, (Class<List<?>>) Collections.EMPTY_LIST.getClass()));
+        Collections.EMPTY_LIST.getClass(),
+        new EmptyListSerializer(fury, (Class<List<?>>) Collections.EMPTY_LIST.getClass()));
     fury.registerSerializer(
-      Collections.emptySortedSet().getClass(),
-      new EmptySortedSetSerializer(
-        fury, (Class<SortedSet<?>>) Collections.emptySortedSet().getClass()));
+        Collections.emptySortedSet().getClass(),
+        new EmptySortedSetSerializer(
+            fury, (Class<SortedSet<?>>) Collections.emptySortedSet().getClass()));
     fury.registerSerializer(
-      Collections.EMPTY_SET.getClass(),
-      new EmptySetSerializer(fury, (Class<Set<?>>) Collections.EMPTY_SET.getClass()));
+        Collections.EMPTY_SET.getClass(),
+        new EmptySetSerializer(fury, (Class<Set<?>>) Collections.EMPTY_SET.getClass()));
     fury.registerSerializer(
-      Collections.singletonList(null).getClass(),
-      new CollectionsSingletonListSerializer(
-        fury, (Class<List<?>>) Collections.singletonList(null).getClass()));
+        Collections.singletonList(null).getClass(),
+        new CollectionsSingletonListSerializer(
+            fury, (Class<List<?>>) Collections.singletonList(null).getClass()));
     fury.registerSerializer(
-      Collections.singleton(null).getClass(),
-      new CollectionsSingletonSetSerializer(
-        fury, (Class<Set<?>>) Collections.singleton(null).getClass()));
+        Collections.singleton(null).getClass(),
+        new CollectionsSingletonSetSerializer(
+            fury, (Class<Set<?>>) Collections.singleton(null).getClass()));
     fury.registerSerializer(
-      ConcurrentSkipListSet.class,
-      new ConcurrentSkipListSetSerializer(fury, ConcurrentSkipListSet.class));
+        ConcurrentSkipListSet.class,
+        new ConcurrentSkipListSetSerializer(fury, ConcurrentSkipListSet.class));
     fury.registerSerializer(Vector.class, new VectorSerializer(fury, Vector.class));
     fury.registerSerializer(ArrayDeque.class, new ArrayDequeSerializer(fury, ArrayDeque.class));
     fury.registerSerializer(BitSet.class, new BitSetSerializer(fury, BitSet.class));
     fury.registerSerializer(
-      PriorityQueue.class, new PriorityQueueSerializer(fury, PriorityQueue.class));
+        PriorityQueue.class, new PriorityQueueSerializer(fury, PriorityQueue.class));
     fury.registerSerializer(
-      CopyOnWriteArrayList.class,
-      new CopyOnWriteArrayListSerializer(fury, CopyOnWriteArrayList.class));
+        CopyOnWriteArrayList.class,
+        new CopyOnWriteArrayListSerializer(fury, CopyOnWriteArrayList.class));
     final Class setFromMapClass = Collections.newSetFromMap(new HashMap<>()).getClass();
     fury.registerSerializer(setFromMapClass, new SetFromMapSerializer(fury, setFromMapClass));
     fury.registerSerializer(
-      ConcurrentHashMap.KeySetView.class,
-      new ConcurrentHashMapKeySetView(fury, ConcurrentHashMap.KeySetView.class));
+        ConcurrentHashMap.KeySetView.class,
+        new ConcurrentHashMapKeySetView(fury, ConcurrentHashMap.KeySetView.class));
   }
 }

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/CollectionSerializers.java
@@ -44,6 +44,7 @@ import java.util.Vector;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentSkipListSet;
 import java.util.concurrent.CopyOnWriteArrayList;
+
 import org.apache.fury.Fury;
 import org.apache.fury.config.Language;
 import org.apache.fury.exception.FuryException;
@@ -51,7 +52,9 @@ import org.apache.fury.memory.MemoryBuffer;
 import org.apache.fury.memory.Platform;
 import org.apache.fury.reflect.ReflectionUtils;
 import org.apache.fury.resolver.ClassInfo;
+import org.apache.fury.resolver.ClassInfoHolder;
 import org.apache.fury.resolver.ClassResolver;
+import org.apache.fury.resolver.RefResolver;
 import org.apache.fury.serializer.ReplaceResolveSerializer;
 import org.apache.fury.serializer.Serializer;
 import org.apache.fury.serializer.Serializers;
@@ -263,7 +266,8 @@ public class CollectionSerializers {
     }
 
     @Override
-    public void write(MemoryBuffer buffer, List<?> value) {}
+    public void write(MemoryBuffer buffer, List<?> value) {
+    }
 
     @Override
     public short getXtypeId() {
@@ -289,7 +293,7 @@ public class CollectionSerializers {
   }
 
   public static class CopyOnWriteArrayListSerializer
-      extends CollectionSerializer<CopyOnWriteArrayList> {
+    extends CollectionSerializer<CopyOnWriteArrayList> {
 
     public CopyOnWriteArrayListSerializer(Fury fury, Class<CopyOnWriteArrayList> type) {
       super(fury, type);
@@ -316,7 +320,8 @@ public class CollectionSerializers {
     }
 
     @Override
-    public void write(MemoryBuffer buffer, Set<?> value) {}
+    public void write(MemoryBuffer buffer, Set<?> value) {
+    }
 
     @Override
     public short getXtypeId() {
@@ -348,7 +353,8 @@ public class CollectionSerializers {
     }
 
     @Override
-    public void write(MemoryBuffer buffer, SortedSet<?> value) {}
+    public void write(MemoryBuffer buffer, SortedSet<?> value) {
+    }
 
     @Override
     public SortedSet<?> read(MemoryBuffer buffer) {
@@ -357,7 +363,7 @@ public class CollectionSerializers {
   }
 
   public static final class CollectionsSingletonListSerializer
-      extends CollectionSerializer<List<?>> {
+    extends CollectionSerializer<List<?>> {
 
     public CollectionsSingletonListSerializer(Fury fury, Class<List<?>> cls) {
       super(fury, cls, false);
@@ -436,7 +442,7 @@ public class CollectionSerializers {
   }
 
   public static final class ConcurrentSkipListSetSerializer
-      extends SortedSetSerializer<ConcurrentSkipListSet> {
+    extends SortedSetSerializer<ConcurrentSkipListSet> {
 
     public ConcurrentSkipListSetSerializer(Fury fury, Class<ConcurrentSkipListSet> cls) {
       super(fury, cls);
@@ -446,9 +452,12 @@ public class CollectionSerializers {
     public ConcurrentSkipListSet newCollection(MemoryBuffer buffer) {
       int numElements = buffer.readVarUint32Small7();
       setNumElements(numElements);
+      RefResolver refResolver = fury.getRefResolver();
+      int refId = refResolver.lastPreservedRefId();
+      // It's possible that comparator/elements has circular ref to set.
       Comparator comparator = (Comparator) fury.readRef(buffer);
       ConcurrentSkipListSet skipListSet = new ConcurrentSkipListSet(comparator);
-      fury.getRefResolver().reference(skipListSet);
+      refResolver.setReadObject(refId, skipListSet);
       return skipListSet;
     }
   }
@@ -473,23 +482,31 @@ public class CollectionSerializers {
     @Override
     public Collection newCollection(MemoryBuffer buffer) {
       final ClassInfo mapClassInfo = fury.getClassResolver().readClassInfo(buffer);
-      final AbstractMapSerializer mapSerializer =
-          (AbstractMapSerializer) fury.getClassResolver().getSerializer(mapClassInfo.getCls());
-      Map map = mapSerializer.newMap(buffer);
-      final int numElements = mapSerializer.getAndClearNumElements();
-      setNumElements(numElements);
-      final Set set = Collections.newSetFromMap(map);
-      fury.getRefResolver().reference(set);
+      final AbstractMapSerializer mapSerializer = (AbstractMapSerializer) mapClassInfo.getSerializer();
+      RefResolver refResolver = fury.getRefResolver();
+      // It's possible that elements or nested fields has circular ref to set.
+      int refId = refResolver.lastPreservedRefId();
+      Map map;
+      if (buffer.readBoolean()) {
+        refResolver.preserveRefId();
+        map = mapSerializer.newMap(buffer);
+        setNumElements(mapSerializer.getAndClearNumElements());
+      } else {
+        map = (Map) mapSerializer.read(buffer);
+        setNumElements(0);
+      }
+      Set set = Collections.newSetFromMap(map);
+      refResolver.setReadObject(refId, set);
       return set;
     }
 
     @Override
     public Collection newCollection(Collection originCollection) {
       Map<?, Boolean> map =
-          (Map<?, Boolean>) Platform.getObject(originCollection, MAP_FIELD_OFFSET);
+        (Map<?, Boolean>) Platform.getObject(originCollection, MAP_FIELD_OFFSET);
       AbstractMapSerializer mapSerializer =
-          (AbstractMapSerializer) fury.getClassResolver().getSerializer(map.getClass());
-      Map newMap = mapSerializer.newMap();
+        (AbstractMapSerializer) fury.getClassResolver().getSerializer(map.getClass());
+      Map newMap = mapSerializer.newMap(map);
       return Collections.newSetFromMap(newMap);
     }
 
@@ -497,32 +514,52 @@ public class CollectionSerializers {
     public Collection onCollectionWrite(MemoryBuffer buffer, Set<?> value) {
       final Map<?, Boolean> map = (Map<?, Boolean>) Platform.getObject(value, MAP_FIELD_OFFSET);
       final ClassInfo classInfo = fury.getClassResolver().getClassInfo(map.getClass());
+      AbstractMapSerializer mapSerializer = (AbstractMapSerializer) classInfo.getSerializer();
       fury.getClassResolver().writeClass(buffer, classInfo);
-      // newMap will read num size first.
-      buffer.writeVarUint32Small7(value.size());
+      if (mapSerializer.supportCodegenHook) {
+        buffer.writeBoolean(true);
+        mapSerializer.onMapWrite(buffer, map);
+      } else {
+        buffer.writeBoolean(false);
+        mapSerializer.write(buffer, map);
+      }
       return value;
     }
   }
 
   public static final class ConcurrentHashMapKeySetView
-      extends CollectionSerializer<ConcurrentHashMap.KeySetView> {
+    extends CollectionSerializer<ConcurrentHashMap.KeySetView> {
+    private final ClassInfoHolder mapClassInfoHolder;
+    private final ClassInfoHolder valueClassInfoHolder;
 
     public ConcurrentHashMapKeySetView(Fury fury, Class<ConcurrentHashMap.KeySetView> type) {
-      super(fury, type);
+      super(fury, type, false);
+      mapClassInfoHolder = fury.getClassResolver().nilClassInfoHolder();
+      valueClassInfoHolder = fury.getClassResolver().nilClassInfoHolder();
     }
 
     @Override
-    public ConcurrentHashMap.KeySetView newCollection(MemoryBuffer buffer) {
-      int numElements = buffer.readVarUint32Small7();
-      setNumElements(numElements);
-      ConcurrentHashMap.KeySetView keySetView = ConcurrentHashMap.newKeySet(numElements);
-      fury.getRefResolver().reference(keySetView);
-      return keySetView;
+    public void write(MemoryBuffer buffer, ConcurrentHashMap.KeySetView value) {
+      fury.writeRef(buffer, value.getMap(), mapClassInfoHolder);
+      fury.writeRef(buffer, value.getMappedValue(), valueClassInfoHolder);
     }
 
     @Override
-    public Collection newCollection(Collection collection) {
-      return ConcurrentHashMap.newKeySet(collection.size());
+    public ConcurrentHashMap.KeySetView read(MemoryBuffer buffer) {
+      ConcurrentHashMap map = (ConcurrentHashMap) fury.readRef(buffer, mapClassInfoHolder);
+      Object value = fury.readRef(buffer, valueClassInfoHolder);
+      return map.keySet(value);
+    }
+
+    @Override
+    public ConcurrentHashMap.KeySetView copy(ConcurrentHashMap.KeySetView value) {
+      ConcurrentHashMap newMap = fury.copyObject(value.getMap());
+      return newMap.keySet(fury.copyObject(value.getMappedValue()));
+    }
+
+    @Override
+    public Collection newCollection(MemoryBuffer buffer) {
+      throw new UnsupportedOperationException();
     }
   }
 
@@ -612,7 +649,7 @@ public class CollectionSerializers {
     public void write(MemoryBuffer buffer, BitSet set) {
       long[] values = set.toLongArray();
       buffer.writePrimitiveArrayWithSize(
-          values, Platform.LONG_ARRAY_OFFSET, Math.multiplyExact(values.length, 8));
+        values, Platform.LONG_ARRAY_OFFSET, Math.multiplyExact(values.length, 8));
     }
 
     @Override
@@ -641,7 +678,7 @@ public class CollectionSerializers {
     @Override
     public Collection newCollection(Collection collection) {
       return new PriorityQueue(
-          collection.size(), fury.copyObject(((PriorityQueue) collection).comparator()));
+        collection.size(), fury.copyObject(((PriorityQueue) collection).comparator()));
     }
 
     @Override
@@ -661,19 +698,19 @@ public class CollectionSerializers {
    * but the correctness can be ensured.
    */
   public static final class DefaultJavaCollectionSerializer<T>
-      extends AbstractCollectionSerializer<T> {
+    extends AbstractCollectionSerializer<T> {
     private Serializer<T> dataSerializer;
 
     public DefaultJavaCollectionSerializer(Fury fury, Class<T> cls) {
       super(fury, cls, false);
       Preconditions.checkArgument(
-          fury.getLanguage() == Language.JAVA,
-          "Python default collection serializer should use " + CollectionSerializer.class);
+        fury.getLanguage() == Language.JAVA,
+        "Python default collection serializer should use " + CollectionSerializer.class);
       fury.getClassResolver().setSerializer(cls, this);
       Class<? extends Serializer> serializerClass =
-          fury.getClassResolver()
-              .getObjectSerializerClass(
-                  cls, sc -> dataSerializer = Serializers.newSerializer(fury, cls, sc));
+        fury.getClassResolver()
+          .getObjectSerializerClass(
+            cls, sc -> dataSerializer = Serializers.newSerializer(fury, cls, sc));
       dataSerializer = Serializers.newSerializer(fury, cls, serializerClass);
       // No need to set object serializer to this, it will be set in class resolver later.
       // fury.getClassResolver().setSerializer(cls, this);
@@ -705,9 +742,11 @@ public class CollectionSerializers {
     }
   }
 
-  /** Collection serializer for class with JDK custom serialization methods defined. */
+  /**
+   * Collection serializer for class with JDK custom serialization methods defined.
+   */
   public static final class JDKCompatibleCollectionSerializer<T>
-      extends AbstractCollectionSerializer<T> {
+    extends AbstractCollectionSerializer<T> {
     private final Serializer serializer;
 
     public JDKCompatibleCollectionSerializer(Fury fury, Class<T> cls) {
@@ -715,9 +754,9 @@ public class CollectionSerializers {
       // Collection which defined `writeReplace` may use this serializer, so check replace/resolve
       // is necessary.
       Class<? extends Serializer> serializerType =
-          ClassResolver.useReplaceResolveSerializer(cls)
-              ? ReplaceResolveSerializer.class
-              : fury.getDefaultJDKStreamSerializerType();
+        ClassResolver.useReplaceResolveSerializer(cls)
+          ? ReplaceResolveSerializer.class
+          : fury.getDefaultJDKStreamSerializerType();
       serializer = Serializers.newSerializer(fury, cls, serializerType);
     }
 
@@ -757,43 +796,43 @@ public class CollectionSerializers {
     Class arrayAsListClass = Arrays.asList(1, 2).getClass();
     fury.registerSerializer(arrayAsListClass, new ArraysAsListSerializer(fury, arrayAsListClass));
     fury.registerSerializer(
-        LinkedList.class, new CollectionSerializer(fury, LinkedList.class, true));
+      LinkedList.class, new CollectionSerializer(fury, LinkedList.class, true));
     fury.registerSerializer(HashSet.class, new HashSetSerializer(fury));
     fury.registerSerializer(LinkedHashSet.class, new LinkedHashSetSerializer(fury));
     fury.registerSerializer(TreeSet.class, new SortedSetSerializer<>(fury, TreeSet.class));
     fury.registerSerializer(
-        Collections.EMPTY_LIST.getClass(),
-        new EmptyListSerializer(fury, (Class<List<?>>) Collections.EMPTY_LIST.getClass()));
+      Collections.EMPTY_LIST.getClass(),
+      new EmptyListSerializer(fury, (Class<List<?>>) Collections.EMPTY_LIST.getClass()));
     fury.registerSerializer(
-        Collections.emptySortedSet().getClass(),
-        new EmptySortedSetSerializer(
-            fury, (Class<SortedSet<?>>) Collections.emptySortedSet().getClass()));
+      Collections.emptySortedSet().getClass(),
+      new EmptySortedSetSerializer(
+        fury, (Class<SortedSet<?>>) Collections.emptySortedSet().getClass()));
     fury.registerSerializer(
-        Collections.EMPTY_SET.getClass(),
-        new EmptySetSerializer(fury, (Class<Set<?>>) Collections.EMPTY_SET.getClass()));
+      Collections.EMPTY_SET.getClass(),
+      new EmptySetSerializer(fury, (Class<Set<?>>) Collections.EMPTY_SET.getClass()));
     fury.registerSerializer(
-        Collections.singletonList(null).getClass(),
-        new CollectionsSingletonListSerializer(
-            fury, (Class<List<?>>) Collections.singletonList(null).getClass()));
+      Collections.singletonList(null).getClass(),
+      new CollectionsSingletonListSerializer(
+        fury, (Class<List<?>>) Collections.singletonList(null).getClass()));
     fury.registerSerializer(
-        Collections.singleton(null).getClass(),
-        new CollectionsSingletonSetSerializer(
-            fury, (Class<Set<?>>) Collections.singleton(null).getClass()));
+      Collections.singleton(null).getClass(),
+      new CollectionsSingletonSetSerializer(
+        fury, (Class<Set<?>>) Collections.singleton(null).getClass()));
     fury.registerSerializer(
-        ConcurrentSkipListSet.class,
-        new ConcurrentSkipListSetSerializer(fury, ConcurrentSkipListSet.class));
+      ConcurrentSkipListSet.class,
+      new ConcurrentSkipListSetSerializer(fury, ConcurrentSkipListSet.class));
     fury.registerSerializer(Vector.class, new VectorSerializer(fury, Vector.class));
     fury.registerSerializer(ArrayDeque.class, new ArrayDequeSerializer(fury, ArrayDeque.class));
     fury.registerSerializer(BitSet.class, new BitSetSerializer(fury, BitSet.class));
     fury.registerSerializer(
-        PriorityQueue.class, new PriorityQueueSerializer(fury, PriorityQueue.class));
+      PriorityQueue.class, new PriorityQueueSerializer(fury, PriorityQueue.class));
     fury.registerSerializer(
-        CopyOnWriteArrayList.class,
-        new CopyOnWriteArrayListSerializer(fury, CopyOnWriteArrayList.class));
+      CopyOnWriteArrayList.class,
+      new CopyOnWriteArrayListSerializer(fury, CopyOnWriteArrayList.class));
     final Class setFromMapClass = Collections.newSetFromMap(new HashMap<>()).getClass();
     fury.registerSerializer(setFromMapClass, new SetFromMapSerializer(fury, setFromMapClass));
     fury.registerSerializer(
-        ConcurrentHashMap.KeySetView.class,
-        new ConcurrentHashMapKeySetView(fury, ConcurrentHashMap.KeySetView.class));
+      ConcurrentHashMap.KeySetView.class,
+      new ConcurrentHashMapKeySetView(fury, ConcurrentHashMap.KeySetView.class));
   }
 }

--- a/java/fury-core/src/test/java/org/apache/fury/FuryTestBase.java
+++ b/java/fury-core/src/test/java/org/apache/fury/FuryTestBase.java
@@ -247,19 +247,19 @@ public abstract class FuryTestBase {
     return (T) fury.deserialize(bytes);
   }
 
-  public static Object serDe(Fury fury1, Fury fury2, Object obj) {
+  public static <T> T serDe(Fury fury1, Fury fury2, T obj) {
     byte[] bytes = fury1.serialize(obj);
-    return fury2.deserialize(bytes);
+    return (T) fury2.deserialize(bytes);
   }
 
-  public static Object serDeCheck(Fury fury1, Fury fury2, Object obj) {
-    Object o = serDe(fury1, fury2, obj);
+  public static <T> T serDeCheck(Fury fury1, Fury fury2, T obj) {
+    T o = serDe(fury1, fury2, obj);
     Assert.assertEquals(o, obj);
     return o;
   }
 
-  public static Object serDeCheck(Fury fury, Object obj) {
-    Object o = serDe(fury, obj);
+  public static <T> T serDeCheck(Fury fury, T obj) {
+    T o = serDe(fury, obj);
     Assert.assertEquals(o, obj);
     return o;
   }

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -436,7 +436,7 @@ public class CollectionSerializersTest extends FuryTestBase {
     Assert.assertEquals(set, serDe(fury, set));
     Assert.assertEquals(
         fury.getClassResolver().getSerializerClass(set.getClass()),
-        CollectionSerializers.ConcurrentHashMapKeySetView.class);
+        CollectionSerializers.ConcurrentHashMapKeySetViewSerializer.class);
     serDeCheck(fury, new CollectionViewTestStruct(set, set));
   }
 

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -435,7 +435,7 @@ public class CollectionSerializersTest extends FuryTestBase {
     set.add("c");
     Assert.assertEquals(set, serDe(fury, set));
     Assert.assertEquals(
-        getJavaFury().getClassResolver().getSerializerClass(set.getClass()),
+        fury.getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.ConcurrentHashMapKeySetView.class);
     serDeCheck(fury, new CollectionViewTestStruct(set, set));
   }

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -410,12 +410,18 @@ public class CollectionSerializersTest extends FuryTestBase {
     Assert.assertEquals(
         fury.getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.SetFromMapSerializer.class);
-    serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    CollectionViewTestStruct struct1 = serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    if (fury.trackingRef()) {
+      assertSame(struct1.collection, struct1.set);
+    }
     set = Collections.newSetFromMap(new HashMap<String, Boolean>() {});
     set.add("a");
     set.add("b");
     serDeCheck(fury, set);
-    serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    CollectionViewTestStruct struct2 = serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    if (fury.trackingRef()) {
+      assertSame(struct2.collection, struct2.set);
+    }
   }
 
   @Test(dataProvider = "furyCopyConfig")
@@ -437,7 +443,10 @@ public class CollectionSerializersTest extends FuryTestBase {
     Assert.assertEquals(
         fury.getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.ConcurrentHashMapKeySetViewSerializer.class);
-    serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    CollectionViewTestStruct o = serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    if (fury.trackingRef()) {
+      assertSame(o.collection, o.set);
+    }
   }
 
   @Test(dataProvider = "furyCopyConfig")

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -408,7 +408,7 @@ public class CollectionSerializersTest extends FuryTestBase {
     set.add("c");
     serDeCheck(fury, set);
     Assert.assertEquals(
-        getJavaFury().getClassResolver().getSerializerClass(set.getClass()),
+        fury.getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.SetFromMapSerializer.class);
     serDeCheck(fury, new CollectionViewTestStruct(set, set));
     set = Collections.newSetFromMap(new HashMap<String, Boolean>() {});

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -38,6 +38,7 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.EnumSet;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashSet;
@@ -401,14 +402,19 @@ public class CollectionSerializersTest extends FuryTestBase {
 
   @Test(dataProvider = "javaFury")
   public void testSetFromMap(Fury fury) {
-    final Set<String> set = Collections.newSetFromMap(Maps.newConcurrentMap());
+    Set<String> set = Collections.newSetFromMap(Maps.newConcurrentMap());
     set.add("a");
     set.add("b");
     set.add("c");
-    Assert.assertEquals(set, serDe(fury, set));
+    serDeCheck(fury, set);
     Assert.assertEquals(
         getJavaFury().getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.SetFromMapSerializer.class);
+    serDeCheck(fury, new CollectionViewTestStruct(set, set));
+    set = Collections.newSetFromMap(new HashMap<String, Boolean>() {});
+    set.add("a");
+    set.add("b");
+    serDeCheck(fury, set);
     serDeCheck(fury, new CollectionViewTestStruct(set, set));
   }
 

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -392,14 +392,15 @@ public class CollectionSerializersTest extends FuryTestBase {
     copyCheck(fury, list);
   }
 
-  @Test(dataProvider = "enableCodegen")
-  public void testSetFromMap(boolean enableCodegen) {
-    final Fury fury =
-        Fury.builder()
-            .withLanguage(Language.JAVA)
-            .requireClassRegistration(false)
-            .withCodegen(enableCodegen)
-            .build();
+  @Data
+  @AllArgsConstructor
+  public static class CollectionViewTestStruct {
+    Collection<String> collection;
+    Set<String> set;
+  }
+
+  @Test(dataProvider = "javaFury")
+  public void testSetFromMap(Fury fury) {
     final Set<String> set = Collections.newSetFromMap(Maps.newConcurrentMap());
     set.add("a");
     set.add("b");
@@ -408,10 +409,11 @@ public class CollectionSerializersTest extends FuryTestBase {
     Assert.assertEquals(
         getJavaFury().getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.SetFromMapSerializer.class);
+    serDeCheck(fury, new CollectionViewTestStruct(set, set));
   }
 
   @Test(dataProvider = "furyCopyConfig")
-  public void testSetFromMap(Fury fury) {
+  public void testSetFromMapCopy(Fury fury) {
     final Set<Object> set = Collections.newSetFromMap(Maps.newConcurrentMap());
     set.add("a");
     set.add("b");
@@ -419,20 +421,21 @@ public class CollectionSerializersTest extends FuryTestBase {
     copyCheck(fury, set);
   }
 
-  @Test
-  public void testConcurrentMapKeySetViewMap() {
-    final ConcurrentHashMap.KeySetView<Object, Boolean> set = ConcurrentHashMap.newKeySet();
+  @Test(dataProvider = "javaFury")
+  public void testConcurrentMapKeySetViewMap(Fury fury) {
+    final ConcurrentHashMap.KeySetView<String, Boolean> set = ConcurrentHashMap.newKeySet();
     set.add("a");
     set.add("b");
     set.add("c");
-    Assert.assertEquals(set, serDe(getJavaFury(), set));
+    Assert.assertEquals(set, serDe(fury, set));
     Assert.assertEquals(
         getJavaFury().getClassResolver().getSerializerClass(set.getClass()),
         CollectionSerializers.ConcurrentHashMapKeySetView.class);
+    serDeCheck(fury, new CollectionViewTestStruct(set, set));
   }
 
   @Test(dataProvider = "furyCopyConfig")
-  public void testConcurrentMapKeySetViewMap(Fury fury) {
+  public void testConcurrentMapKeySetViewMapCopy(Fury fury) {
     final ConcurrentHashMap.KeySetView<Object, Boolean> set = ConcurrentHashMap.newKeySet();
     set.add("a");
     set.add("b");

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -73,6 +73,7 @@ import org.testng.Assert;
 import org.testng.annotations.Test;
 import org.testng.collections.Maps;
 
+@SuppressWarnings("rawtypes")
 public class CollectionSerializersTest extends FuryTestBase {
 
   @Test(dataProvider = "referenceTrackingConfig")
@@ -446,6 +447,12 @@ public class CollectionSerializersTest extends FuryTestBase {
     CollectionViewTestStruct o = serDeCheck(fury, new CollectionViewTestStruct(set, set));
     if (fury.trackingRef()) {
       assertSame(o.collection, o.set);
+    }
+    ConcurrentHashMap<String, String> map = new ConcurrentHashMap<>();
+    map.put("k1", "v1");
+    if (fury.trackingRef()) {
+      ArrayList<Serializable> list = serDeCheck(fury, ofArrayList(map.keySet("v0"), map));
+      assertSame(((ConcurrentHashMap.KeySetView) (list.get(0))).getMap(), list.get(1));
     }
   }
 

